numpy/_core/src/multiarray/iterators.c

```diff
@@ -857,201 +811,154 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
+
+    PyArray_Descr *dtype = PyArray_DESCR(self->ao);
+    npy_intp dtype_size = dtype->elsize;
+    NPY_cast_info cast_info = {.func = NULL};
+
+    /* Prepare the indices */
+    index_type = prepare_index_noarray(1, &self->size, ind, indices, &index_num,
+        &ndim, &fancy_ndim, 1, 1);
+
+    if (index_type < 0) {
+        goto finish;
+    }
+
+    else if (indices[0].type == HAS_NEWAXIS) {
+        PyErr_SetString(PyExc_IndexError,
+            "only integers, slices (`:`), ellipsis (`...`) and integer or boolean "
+            "arrays are valid indices"
+        );
+        goto finish;
     }
 
-    if (PyTuple_Check(ind)) {
-        int len;
-        len = PyTuple_GET_SIZE(ind);
-        if (len > 1) {
+    // Single ellipsis index
+    else if (index_type == HAS_ELLIPSIS) {
+        ind = PySlice_New(NULL, NULL, NULL);
+        if (ind == NULL) {
             goto finish;
         }
-        ind = PyTuple_GET_ITEM(ind, 0);
-    }
 
-    type = PyArray_DESCR(self->ao);
+        ret = iter_ass_subscript(self, ind, val);
+        Py_DECREF(ind);
+        goto finish;
+    }
 
-    /*
-     * Check for Boolean -- this is first because
-     * Bool is a subclass of Int
-     */
-   
-    if (PyBool_Check(ind)) {
-        retval = 0;
-        int istrue = PyObject_IsTrue(ind);
-        if (istrue == -1) {
-            return -1;
-        }
-        if (istrue) {
-            retval = PyArray_Pack(
-                    PyArray_DESCR(self->ao), self->dataptr, val);
+    // Single boolean index
+    else if (indices[0].type == HAS_0D_BOOL) {
+        ret = 0;
+        if (indices[0].value) {
+            ret = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
         }
         goto finish;
     }
 
-    if (PySequence_Check(ind) || PySlice_Check(ind)) {
-        goto skip;
-    }
-    start = PyArray_PyIntAsIntp(ind);
-    if (error_converting(start)) {
-        PyErr_Clear();
-    }
-    else {
-        if (check_and_adjust_index(&start, self->size, -1, NULL) < 0) {
+    PyArray_ITER_RESET(self);
+
+    if (index_type == HAS_INTEGER) {
+        if (check_and_adjust_index(&indices[0].value, self->size, -1, NULL) < 0) {
             goto finish;
         }
-        PyArray_ITER_GOTO1D(self, start);
-        retval = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
+        PyArray_ITER_GOTO1D(self, indices[0].value);
+        ret = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
         PyArray_ITER_RESET(self);
-        if (retval < 0) {
+        if (ret < 0) {
             PyErr_SetString(PyExc_ValueError,
-                            "Error setting single item of array.");
+                            "Error setting single item of flat iterator.");
         }
         goto finish;
     }
 
- skip:
-    Py_INCREF(type);
-    arrval = (PyArrayObject *)PyArray_FromAny(val, type, 0, 0,
+    Py_INCREF(dtype);
+    PyArrayObject *arrval = (PyArrayObject *)PyArray_FromAny(val, dtype, 0, 0,
```

Comment:
Would be good to pass the correct maxdims here, IIRC (fixes corner cases around object arrays, even if the choice of how that behaves is a matter of taste).

Reply:
How would we do that? `array_prepare_index` uses this exact construct.

Reply:
The `0, 0` would be `0, maxdims`, but matching the normal indexing makes sense indeed.